Mathlib/Data/ListM.lean

```diff
@@ -0,0 +1,268 @@
+/-
+Copyright (c) 2018 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Keeley Hoek, Simon Hudon, Scott Morrison
+
+! This file was ported from Lean 3 source module data.mllist
+! leanprover-community/mathlib commit 8f6fd1b69096c6a587f745d354306c0d46396915
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Option.Defs
+
+/-! # Monadic lazy lists.
+
+Lazy lists with "lazyness" controlled by an arbitrary monad.
+
+The inductive construction is unsafe (indeed, we can build infinite objects).
+This isn't so bad, as the typical use is with the tactic monad, in any case.
+Typical usage will be to do computations in `ListM`, and then eventually extract
+the first result (with `head` or `firstM`), first several results (using `take`),
+or all results (using `force`). At this point you can use the `unsafe` term elaborator
+from `Std.Util.TermUnsafe` to return to the "safe" world.
+
+As we're unsafe anyway, we don't bother with proofs about these constructions.
+-/
+
+universe u v
+
+/-- A monadic lazy list, controlled by an arbitrary monad. -/
+unsafe inductive ListM (m : Type u → Type u) (α : Type u) : Type u
+  | nil : ListM m α
+  | cons : m (Option α × ListM m α) → ListM m α
+#align tactic.mllist ListM
```

Comment:
Did we drop the extra `l` (lazy) deliberately?

Reply:
Arguably whether it is lazy depends on which monad you put here.